File: 2023/javascript/01/01.js
export function first(input) {
  const lines = input.split("\n").filter(Boolean);

  const nums = lines.map((line) =>
    line
      .split(/[a-zA-Z]/)
      .filter(Boolean)
      .join("")
      .split("")
  );

  return nums.reduce((prev, curr) => {
    const str = `${curr[0]}${curr[curr.length - 1]}`;
    return prev + Number(str);
  }, 0);
}

const numTuple = [
  ["0", "zero"],
  ["1", "one"],
  ["2", "two"],
  ["3", "three"],
  ["4", "four"],
  ["5", "five"],
  ["6", "six"],
  ["7", "seven"],
  ["8", "eight"],
  ["9", "nine"],
];

const numsMap = numTuple.reduce((prev, [a, b], i) => {
  return {
    ...prev,
    [a]: i,
    [b]: i,
  };
}, {});

const chars = numTuple
  .map(([a, b]) => {
    return `${a}|${b}`;
  }, "")
  .join("|");

const regex = new RegExp(`(?=(${chars}))`, "g");

export function second(input) {
  const lines = input.split("\n").filter(Boolean);

  const numPairs = lines.map((s) => {
    const nums = [];

    let m = regex.exec(s);
    while (m !== null) {
      if (m.index === regex.lastIndex) {
        regex.lastIndex++;
      }

      m.forEach((match) => {
        if (match) nums.push(numsMap[match]);
      });

      m = regex.exec(s);
    }

    return nums;
  });

  return numPairs.reduce((prev, curr) => {
    const str = `${curr[0]}${curr[curr.length - 1]}`;
    return prev + Number(str);
  }, 0);
}

File: 2023/go/01.go
package advent2023

import (
	"slices"
	"strconv"
	"strings"
)

func Calibrate(input string) int {
	lines := strings.Split(input, "\n")
	sum := 0
	for _, line := range lines {
		if line == "" {
			continue
		}

		sum += returnFirstLast(line)
	}

	return sum
}

var words = []string{"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}

func CalibrateWithWords(input string) int {
	lines := strings.Split(input, "\n")
	sum := 0

	for _, line := range lines {
		if line == "" {
			continue
		}

		first := findFirstMatch(words, line)
		last := findLastMatch(words, line)

		firstIndex := slices.Index(words, first)
		lastIndex := slices.Index(words, last)

		sum += firstIndex + lastIndex
	}
	return sum
}

func returnFirstLast(line string) int {
	firstChar := ""
	for _, char := range line {
		if char >= '0' && char <= '9' {
			firstChar = string(char)
			break
		}
	}

	lastChar := ""
	for i := len(line) - 1; i >= 0; i-- {
		if line[i] >= '0' && line[i] <= '9' {
			lastChar = string(line[i])
			break
		}
	}

	num, _ := strconv.Atoi(firstChar + lastChar)

	return num
}

func findFirstMatch(strs []string, s string) string {
	minIndex := -1
	var result string

	for _, substr := range strs {
		index := strings.Index(s, substr)
		if index >= 0 {
			if minIndex == -1 || index > index {
				minIndex = index
				result = substr
			}
		}
	}

	return result
}

func findLastMatch(strs []string, s string) string {
	maxIndex := -1
	var result string

	for _, substr := range strs {
		index := strings.LastIndex(s, substr)
		if index > maxIndex {
			maxIndex = index
			result = substr
		}
	}

	return result
}
